pkg/dynamiclistener/server.go

```diff
@@ -95,9 +95,11 @@ func (s *Server) updateIPs(savedIPs map[string]bool) map[string]bool {
 	}
 
 	certs := map[string]string{}
+	s.Lock()
 	for key, cert := range s.certs {
 		certs[key] = certToString(cert)
 	}
+	s.Unlock()
```

Comment:
is it possible to hit a similar bug at line 114?
```
	for _, k := range s.ips.Keys() {
```
should we lock arond that too?

Reply:
The `ips` LRU contains a `sync.RWMutex` and code in `Keys()` to `RLock` and `RUnlock`. That mutex combined with the fact that we aren't modifying `ips` in the loop means that we should not need to server lock around `range s.ips.Keys()`.